--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -26,6 +26,23 @@
 #include <linux/bootmem.h>
 #include <linux/task_work.h>
 #include <linux/sched/task.h>
 #include <linux/fs_context.h>
+#include <linux/susfs.h>
+
+#ifndef DEFAULT_KSU_MNT_ID
+#define DEFAULT_KSU_MNT_ID 1000000000
+#endif
+#ifndef DEFAULT_KSU_MNT_GROUP_ID
+#define DEFAULT_KSU_MNT_GROUP_ID 1000000000
+#endif
+#ifndef VFSMOUNT_MNT_FLAGS_KSU_UNSHARED_MNT
+#define VFSMOUNT_MNT_FLAGS_KSU_UNSHARED_MNT 0x2000000
+#endif
+#ifndef CL_COPY_MNT_NS
+#define CL_COPY_MNT_NS (1<<23)
+#endif
+
+DEFINE_IDA(susfs_mnt_id_ida);
+DEFINE_IDA(susfs_mnt_group_ida);
+bool susfs_is_sdcard_android_data_decrypted = false;
+extern int susfs_is_current_ksu_domain(void);
 
 #include "pnode.h"
 #include "internal.h"
